### DISCUSSION TITLE
EFX expression context vs. xpath context

### DISCUSSION THREAD

**Discussion by karel-debrabandere-bosa-fgov-be** [OP-TED/eForms-SDK#228] [Q&A] [answered]
Every EFX expression in the SDK is structured in a way that it implies in the context of which conceptual node it should be evaluated (`{ND-Root} ${ ... }`). When using the EFX toolkit (provided by TED) to convert these expressions into xpath 2.0 expressions, this context is lost.

When analysing this in the past, we saw that in every case, the context specified was actually the conceptual parent of the current field node. We use that rule for evaluating our xpath expressions.

However, when looking into implementing the `assert` feature (which I believe was introduced in eforms 1.3.0?),  we see that this rule no longer seems to apply: the conditions in the assert constraints (as well as the constraint value) specify a conceptual context that is not necessarily the node's parent. (NB: initial analysis suggests that it is some higher ancestor). Because the EFX evaluation context is not part of the matching xpath expression, there is currenly no reliable way for our eforms engine to know on which context to evaluate these expressions.

**Comment by YvesJo** [comment, marked as answer]
Indeed, the conceptual parent node was initially foreseen as a context for the rules, however requests were made to adapt this in order to simplify the XML generation and the parent node had to be adapted for some fields, requiring the specification of the rule context at the rule level when it differed from the parent node of the field.
